### ISSUE TITLE
Github API rate limit being triggered by automatic upgrades

### ISSUE THREAD

**Issue by alyandon** [syncthing/syncthing#9033]
Versions:  Various - as old as 1.25.0 to current
OS: Windows

Problem:  This has been an ongoing issue for me where I have multiple Windows machines that refuse to automatically update which forces me to download the latest binary manually and copy it into my syncthing directory.  They error is always "Automatic upgrade: zip: not a valid zip file".  I disabled Windows Defender as a first troubleshooting step but that did not resolve the issue.

I decided to take it upon myself to add some debugging code and build an older version of syncthing so the automatic upgrade triggers when I start it.  Here is an excerpt from the result:

```
[IDMZY] 2023/08/09 08:44:00 INFO: Automatic upgrade (current "v1.23.6-dirty" < latest "v1.23.7")
### start my logging changes
[IDMZY] 2023/08/09 08:44:00 INFO: loading "https://api.github.com/repos/syncthing/syncthing/releases/assets/120749509"
[IDMZY] 2023/08/09 08:44:00 INFO: readZip is short debug: 279 bytes
[IDMZY] 2023/08/09 08:44:00 INFO: dump: {"message":"API rate limit exceeded for 24.243.x.x. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
### end my logging changes
[IDMZY] 2023/08/09 08:44:00 WARNING: Automatic upgrade: zip: not a valid zip file
```

So, it looks like a rate limit error is being returned by Github when syncthing attempts the fetch but the relevant code is assuming the result is always a valid zip file.

The problem(s) as I see it:

1) I have a moderately large swarm with 5 machines behind that 24.243.x.x address.  Syncthing appears to aggressively query that API endpoint repeatedly every time it connects to a peer and the peer reports it is running a later version.
2) "not a valid zip file" is not really a useful error to report to the user.  Syncthing should verify that the result of that fetch is actually a zip file before feeding it to readZip.

What happened:
Automatic upgrades don't work reliably for me.

What should happen:
Automatic upgrades should Just Work(tm).

**Comment by alyandon** [comment]
Unsurprisingly, I also have a few Linux instances behind that same IP that report:
` WARNING: Automatic upgrade: gzip: invalid header`

I don't usually notice those since they are headless instances.

**Comment by tomasz1986** [comment]
The limit is 60 requests per hour for unauthenticated access. This shouldn't be a problem when using just Syncthing (unless you've got 60+ devices that all check for updates at the same time…). Do you (or anyone else under the same IP address) use other software that also utilises the GitHub API?

**Comment by alyandon** [comment]
> The limit is 60 requests per hour for unauthenticated access. This shouldn't be a problem when using just Syncthing (unless you've got 60+ devices that all check for updates at the same time…). Do you (or anyone else under the same IP address) use other software that also utilises the GitHub API?

I have 7 instances total behind that IP and it appears that syncthing is making a request to api.github.com every time it connects to a peer and that peer reports a later version than the local version which quickly burns through that 60 limit.

I am also a pretend developer by trade so it is entirely possible that other things running on my home network besides syncthing are trying to utilize the Github API as a makeshift CDN and are contributing to that.  However, syncthing is the only thing I am currently aware of that is causing this problem.

**Comment by alyandon** [comment]
If I spent some time to work up a PR that makes syncthing pay attention to the http response code and honor the x-ratelimit headers would it be accepted?

**Comment by alyandon** [comment]
Some more information, this is a single instance of syncthing v1.23.6 running on a linux host that was repeatedly hitting the api in response to remote peers reporting they were on the newer v1.23.7 version:

```Aug 09 06:13:32 nebula rc.local[1078]: [DXNJM] 2023/08/09 06:13:32 WARNING: Automatic upgrade: gzip: invalid header
Aug 09 06:13:35 nebula rc.local[1078]: [DXNJM] 2023/08/09 06:13:35 WARNING: Automatic upgrade: gzip: invalid header
Aug 09 06:14:43 nebula rc.local[1078]: [DXNJM] 2023/08/09 06:14:43 WARNING: Automatic upgrade: gzip: invalid header
Aug 09 06:14:46 nebula rc.local[1078]: [DXNJM] 2023/08/09 06:14:46 WARNING: Automatic upgrade: gzip: invalid header
Aug 09 06:18:43 nebula rc.local[1078]: [DXNJM] 2023/08/09 06:18:43 WARNING: Automatic upgrade: gzip: invalid header
Aug 09 06:20:56 nebula rc.local[1078]: [DXNJM] 2023/08/09 06:20:56 WARNING: Automatic upgrade: gzip: invalid header
Aug 09 06:21:17 nebula rc.local[1078]: [DXNJM] 2023/08/09 06:21:17 WARNING: Automatic upgrade: gzip: invalid header```

**Comment by calmh** [comment]
To be clear, this is not the upgrade check, it's the actual upgrade download. The upgrade check, which happens somewhat frequently (for example every time a newer machine connects to us) goes to upgrades.syncthing.net which is our own server, with caching, entirely sidestepping the github API limits.

However, when an upgrade is deemed possible and necessary the asset is downloaded from the GitHub release, which is what happens here (and gets rejected). I suspect think that's a fairly unusual situation, as it is literally one "API request" (actually zip/tar.gz download) per released version. I guess if you're already at or above the GitHub API limit for other reasons this will be one more thing that fails.

**Comment by alyandon** [comment]
@calmh Thanks for the clarification.  However, after digging into this and looking at network traffic I have not identified anything else that would be interacting with api.github.com repeatedly and exhausting that rate limit that isn't traceable to syncthing.  I agree that there is a non-zero chance that syncthing is an innocent bystander but I haven't identified another culprit yet and I don't run any software that utilizes the GH API that I'm aware of.

Also, it's not a cluster of 5 machines - it is a cluster of about 10-15 or so devices (not all of which are continuously connected) with only half of them behind this NAT gateway.  I'm not sure if that changes the situation at all.

Regardless, no matter what is going on I feel that checking the http status code and honoring the x-rate-limit flag should be a good improvement to consider.

**Comment by bt90** [comment]
Is Githubs container registry subject to the same ratelimit?

**Comment by alyandon** [comment]
Actually, I may have a lead on another culprit - was looking for traffic sent to 140.82.114.3 when I really needed to be looking at 140.82/16.

**Comment by alyandon** [comment]
Alright, for future generations that come after me.  I have a Raspberry Pi that I send to people to help diagnose network issues remotely so they can hold their ISPs accountable.  Part of the software stack running on it includes telegraf which is configured to collect tcp/icmp response telemetry against a variety of public and semi-public remote hosts.

One of those public hosts happens to be `api.github.com` and it appears that even a simple HEAD request to https://api.github.com/ counts against the hourly quota now.  So - I think it is safe to say that syncthing was (mostly) an innocent bystander but it does contribute to the problem by hosting assets in GH.

So, while technically not-a-bug - I think syncthing's behavior could be improved by:

1) paying attention to http status codes when fetching remote resources and logging a meaningful error message on failure
2) if a rate limit response is returned, honor the x-ratelimit-rest header instead of continuing to make requests and log the message so it is clear to the user what is happening

Potential feature request?  Should I go post this in the forum?
1) Provide a mechanism for configuring and using personal GH tokens to fetch GH resources since they have a much more generous rate limit

**Comment by bt90** [comment]
The status should be properly logged. But i think the retry-after handling is a bit of an overkill here.

**Comment by alyandon** [comment]
I won't pretend to have fully internalized syncthing's code base at this point but is it really overkill to have a map with a host and retry-after entry in it to check before making the http call?  I would think that honoring the remote host's request that the client wait to retry the call is being a good netizen but maybe my scenario was enough of a corner case that it isn't worth the effort.

**Comment by calmh** [comment]
I think this can be resolved more simply by resolving the URL used one step further, IIRC the url seem above mostly results in a redirect, which we could do already in the upgrade server before serving the response.

**Comment by calmh** [comment]
There, no requests towards the api.github.com domain. I think that should solve it.

**Comment by alyandon** [comment]
Indeed.  Thanks for all the effort you and everyone else put into this project.
